internal/provider/oresource/base_resource.go

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package oresource
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+
+	"terraform-provider-oodle/internal/oodlehttp"
+	"terraform-provider-oodle/internal/oodlehttp/clientmodels"
+	"terraform-provider-oodle/internal/resourceutils"
+)
+
+type BaseResource[M clientmodels.ClientModel, R resourceutils.ResourceModel[M]] struct {
+	client           *oodlehttp.ModelClient[M]
+	newResourceModel func() R
+	newClientModel   func() M
+	clientCreator    func(oodleHttpClient *oodlehttp.OodleApiClient) *oodlehttp.ModelClient[M]
```

Comment:
better to name functions that read nice at the call site more than at the def site

at call site, this reads `foo.clientCreator(args)`
`foo.createClient(args)` or `foo.newClient(args)` would be better